Embed or attach files based on their file extension

dzsibi opened the pull request:
There have been a few discussions about embedding images in e-mails (#251, #884). As it is right now, while Content-IDs were added, the Content-Disposition header was still set to attachment for all files, and as such, e-mail clients showed them as attached files. My proposed solution makes the judgement on whether a file should be embedded or attached based on its extension - jpg, jpeg, png and gif files are embedded (and have a Content-ID set), while other file types are attached. 

It might be better on the long run to offer a checkbox to set whether a file should be attached or embedded, but I believe that 99% of the time images are only added to a message to be shown inline. With these changes, adding images to a message will work as expected, while other files can still be added as attachments.

dzsibi committed:
Embed or attach files based on their file extension

Stuw commented:
Tested - patch works well.

I've applied it on top of master and tested email template with docx and jpg.
Image was embedded as expected, document was attached.

Thanks for the patch!

jordan-wright commented:
Hey there!

Thanks for sending this in- sorry for the delay in getting back to you. Can you expand a bit on this workflow? How are you adding images in the editor?

I’m excited for this since it definitely is a pain point for folks. Thanks for taking this on!

dzsibi commented:
Hey!

So the way this works is that once you add an image with the correct extension, you can then reference it in image tags or CSS using its name. You have to use the "cid:" URL format, like this:

`<img src="cid:image.png" />`

Now this can only be done using the source code editor (I don't really use the WYSIWYG editor), but I guess in theory one could create data URLs from the uploaded images and use them to stand in place for the CID URLs when displaying the images in the editor.

dzsibi commented:
Hey, 

So I saw the discussion in #1865, and I would like to clarify a few points:

- Referencing images by CID will work without this patch
- The images will still show up as attachments without this patch

The reason for this is that when Gophish [uses](https://github.com/gophish/gophish/blob/ec8b17238ee725e84f0d2cff4c20c2b8bbb362f5/models/email_request.go#L172) msg.Attach to add the image files, the Content-Disposition header will be set to attachment instead of inline, this can be traced to gomail's implementation, see [here](https://github.com/go-gomail/gomail/blob/81ebce5c23dfd25c6c67194b37d3dd3f338c98b1/writeto.go#L131).

glennzw commented:
@dzsibi Yeah I use your patch in production, it works great, thank you.

nerotix commented:
I really this change, how would I go to try and get this running in my local installation? I tried building from source and make your changes inside the files and build it but that destroys the whole email templates section within the webgui.

fumamatar commented:
Are there any reasons why this is not merged? Inline attachments are vital to our use case and I am sure to many others too.

glennzw commented:
Hi @dzsibi. Thanks so much for your work on this, and your patience with us getting it merged in. I'm running some tests, and have noticed that all attachments get marked with `Content-Disposition: inline` (not just the image extensions in your `EmbeddedFileExtensions` slice. e.g here's a .docx attachment:

<img width="884" alt="Screenshot 2021-12-22 at 18 22 08" src="https://user-images.githubusercontent.com/3966613/147132102-e58a292d-9c0b-465e-a287-07fd49202599.png">


I tested master vs your PR with two images and a docx, these are the results:

Running on master I get:
```
Content-Disposition: attachment; filename="cat.docx"
Content-Disposition: attachment; filename="cat.jpg"
Content-Disposition: attachment; filename="cat.png"
```

Running on your PR I get:
```
Content-Disposition: inline; filename="cat.docx"
Content-Disposition: inline; filename="cat.jpg"
Content-Disposition: inline; filename="cat.png"

```

glennzw commented:
The problem seems to be with this logic:

```go
		if sort.SearchStrings(EmbeddedFileExtensions, filepath.Ext(a.Name)) < len(EmbeddedFileExtensions) {
			msg.Embed(a.Name, copyFunc)
		} else {
			msg.Attach(a.Name, copyFunc)
		}
```
The _if statement_ always returns True, because if an extension is not in the list, an index position of zero is returned, which is less than `len(EmbeddedFileExtensions)`. The return value is the index to insert x if x is not present (so if not present it will return 0, because that's where you should insert it).

Also, `SearchStrings` [expects](https://pkg.go.dev/sort) a sorted slice, and your slice is not sorted so searching for, example, ".gif" returns an index of 0 (should return 4):

```go
var EmbeddedFileExtensions = []string{".jpg", ".jpeg", ".png", ".gif"}
```

So I don't think `sort.SearchStrings` is what we want to use. How about this:

```go
func contains(s []string, searchterm string) bool {
	i := sort.SearchStrings(s, searchterm)
	return i < len(s) && s[i] == searchterm
}
```

and then:

```go
                if contains(EmbeddedFileExtensions, filepath.Ext(a.Name)) {
			msg.Embed(a.Name, copyFunc)
		} else {
			msg.Attach(a.Name, copyFunc)
		}
```

This seems to work, you can test it here: https://go.dev/play/p/H7oSyscky1T

dzsibi committed:
Fix extension check for embedded file types, add tests for embedding

dzsibi committed:
Remove duplicate code when adding attachments

dzsibi committed:
Restore attachment template support

dzsibi commented:
Hey [glennzw](https://github.com/glennzw),

I rebased against the latest master and fixed the issue you mentioned. It was truly a rookie mistake, my apologies. Also added some tests just to make sure it works as expected. I also cleaned up the merge conflict with the templating support for attachments that was added after the original PR. 

(There are some formatting errors on the current master that are unrelated to my code. I fixed it on my own branch, but didn't want to polute the PR with these unrelated changes.)

glennzw commented:
Hi @dzsibi, thanks so much for following up with this, I think it's a great feature to add! No worries on the mistake, I'm quite the expert at making them. I'll take this for a spin now, really appreciate you adding the test conditions and rebasing.

Good spot on the formatting issues, I'd better get on top of that.
